src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -0,0 +1,74 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include <string>
+#include <vector>
+#include <utility>
+#include <stdexcept>
+#include <iostream>
+#include <memory>
+#include <variant>
+#include <optional>
+
+
+
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/ast_construction.h"
+#include "src/ir/datalog/program.h"
+#include "src/common/testing/gtest.h"
+
+namespace raksha::ir::auth_logic {
+
+
+TEST(AstConstructionSuite, SimpleQueryTest) {
+    Parser parser;
+    absl::string_view input_text = R"qql(may_notifierA = query "UserA" says may("Notifier"))qql";
+
+    absl::StatusOr<Program> program_or_null = parser.ParseProgram(input_text);
+    ASSERT_EQ(program_or_null.ok(), true);
+
+    Program prog_out = *std::move(program_or_null);
+    std::vector<std::string> actual_name;
+    std::vector<std::string> actual_principal_name;
+    std::vector<std::string> actual_predicate_name;
+    std::vector<std::string> actual_predicate_args;
+
+    for ( auto const &each_query : prog_out.queries()) {
+         actual_name.push_back(each_query.name()); 
+         actual_principal_name.push_back(each_query.principal().name()); 
+         BaseFact basefact(std::move(std::get<BaseFact>(each_query.fact().GetValue())));
+         actual_predicate_name.push_back(std::get<datalog::Predicate>(basefact.GetValue()).name());
+         for (auto const &each_arg : std::get<datalog::Predicate>(basefact.GetValue()).args() ) {
+            actual_predicate_args.push_back(each_arg);
+         }
+        
+     } 
+
+    std::string expected_name = "may_notifierA";
+    std::string expected_principal_name = "\"UserA\"";
+    std::string expected_predicate_name = "may";
+    std::string expected_predicate_args = "\"Notifier\"";
+
+    EXPECT_EQ(actual_name[0], expected_name);
+    EXPECT_EQ(actual_principal_name[0], expected_principal_name);
+    EXPECT_EQ(actual_predicate_name[0], expected_predicate_name);
+    EXPECT_EQ(actual_predicate_args[0], expected_predicate_args);
```

Comment:
Don't use `operator[]` on `std::vector`s. It performs a non-bounds-checked access and can overflow or underflow. The `at` method does the same thing, but is safe.